File: frontend/lib/api/activity.ts
"use client";

import { apiFetch } from "./client";

export type ActivitySubjectType = "file" | "folder" | "tag" | "storage" | "share";

export interface ActivityEntry {
  id: string;
  subject_type: ActivitySubjectType;
  subject_id: string;
  action: string;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export const activityApi = {
  listForSubject: (subjectID: string, limit = 50) => {
    const qs = new URLSearchParams({
      subject_id: subjectID,
      limit: String(limit),
    }).toString();
    return apiFetch<ActivityEntry[]>(`/activity?${qs}`);
  },
  recent: (limit = 20) => {
    const qs = new URLSearchParams({ limit: String(limit) }).toString();
    return apiFetch<ActivityEntry[]>(`/activity/recent?${qs}`);
  },
};

File: backend/internal/domain/activity.go
package domain

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ActivitySubjectType string

const (
	ActivitySubjectFile    ActivitySubjectType = "file"
	ActivitySubjectFolder  ActivitySubjectType = "folder"
	ActivitySubjectTag     ActivitySubjectType = "tag"
	ActivitySubjectStorage ActivitySubjectType = "storage"
	ActivitySubjectShare   ActivitySubjectType = "share"
)

// ActivityAction is a free-form verb such as "upload_complete", "delete",
// "restore", "rename", "move", "create", "revoke".
type ActivityAction string

type ActivityLog struct {
	ID          primitive.ObjectID  `bson:"_id,omitempty"`
	UserID      primitive.ObjectID  `bson:"user_id"`
	SubjectType ActivitySubjectType `bson:"subject_type"`
	SubjectID   primitive.ObjectID  `bson:"subject_id"`
	Action      ActivityAction      `bson:"action"`
	Metadata    map[string]any      `bson:"metadata,omitempty"`
	IP          string              `bson:"ip,omitempty"`
	UserAgent   string              `bson:"user_agent,omitempty"`
	CreatedAt   time.Time           `bson:"created_at"`
}
